src/platform/intel/cavs/lib/clk.c

```diff
@@ -40,14 +43,74 @@ static int clock_platform_set_cpu_freq(int clock, int freq_idx)
 	io_reg_update_bits(SHIM_BASE + SHIM_CLKCTL,
 			   SHIM_CLKCTL_OSC_SOURCE_MASK, enc);
 
-	/* release other clocks */
-	io_reg_write(SHIM_BASE + SHIM_CLKCTL,
-		     (io_reg_read(SHIM_BASE + SHIM_CLKCTL) &
-		     ~SHIM_CLKCTL_OSC_REQUEST_MASK) | enc);
+	if (release_unused) {
+		/* release other clocks */
+		io_reg_write(SHIM_BASE + SHIM_CLKCTL,
+			     (io_reg_read(SHIM_BASE + SHIM_CLKCTL) &
+			      ~SHIM_CLKCTL_OSC_REQUEST_MASK) | enc);
+	}
+}
 #endif
+
+static int clock_platform_set_cpu_freq(int clock, int freq_idx)
+{
+	select_cpu_clock(freq_idx, true);
 	return 0;
 }
 
+#if CONFIG_CAVS_USE_LPRO_IN_WAITI
+/* Store clock source that was active before going to waiti,
+ * so it can be restored on wake up.
+ */
+static SHARED_DATA int active_freq_idx = CPU_DEFAULT_IDX;
+
+static inline int get_cpu_current_freq_idx(void)
+{
+	struct clock_info *clk_info = clocks_get() + CLK_CPU(cpu_get_id());
+
+	return clk_info->current_freq_idx;
+}
+
+static inline void set_cpu_current_freq_idx(int freq_idx)
+{
+	int i;
+	struct clock_info *clk_info = clocks_get();
+
+	for (i = 0; i < PLATFORM_CORE_COUNT; i++)
+		clk_info[i].current_freq_idx = freq_idx;
+
+	platform_shared_commit(clk_info,
+			       sizeof(*clk_info) * PLATFORM_CORE_COUNT);
+}
+
+void platform_clock_on_wakeup(void)
+{
+	int freq_idx = *cache_to_uncache(&active_freq_idx);
+
+	if (freq_idx != get_cpu_current_freq_idx()) {
+		select_cpu_clock(freq_idx, true);
+		set_cpu_current_freq_idx(freq_idx);
+	}
+}
+
+void platform_clock_on_waiti(void)
+{
+	int freq_idx = get_cpu_current_freq_idx();
+
+	*cache_to_uncache(&active_freq_idx) = freq_idx;
```

Comment:
I think @tlauda is about to remove all these calls, so maybe best to update.

Reply:
@lgirdwood They were only removed from non-CAVS platforms.

Reply:
@lgirdwood It's going to be removed from non-cavs platforms as @paulstelian97 wrote. This code is cavs-specific so it can use cavs-only functions, that is necessary here cos we deal with shared data.